src/server/engine_shard_set.cc

```diff
@@ -56,6 +71,67 @@ EngineShard::Stats& EngineShard::Stats::operator+=(const EngineShard::Stats& o)
   return *this;
 }
 
+// This function checks 3 things:
+// 1. Don't try memory fragmentation if we don't use "enough" memory (control by
+// mem_defrag_threshold flag)
+// 2. That we had change in memory usage - to prevent endless loop - out of scope for now
+// 3. in case the above is OK, make sure that we have a "gap" between usage and commited memory
+// (control by commit_use_threshold flag)
+bool EngineShard::DefragTaskArgs::IsRequired() {
+  static const uint64_t mem_size = max_memory_limit;
+  static const float threshold_mem = mem_size / GetFlag(FLAGS_mem_defrag_threshold);
+  static const float commit_use_threshold = GetFlag(FLAGS_commit_use_threshold);
+
+  if (cursor > 0) {
+    return true;
+  }
```

Comment:
I think that it will end up been the same. 
I use the first solution first, but it turn out that the cursor > 0 will win anyways. So this function just simplify it, I want to know whether to run or not. If I'm in the middle, then just run, don't care about the other condition, otherwise (we are not in the middle of the scan), check memory usage and do whatever is needed. 
I just want to make the "outside" code simpler in this case.